prime-router/src/main/kotlin/fhirengine/translation/hl7/HL7Truncator.kt

```diff
@@ -0,0 +1,122 @@
+package gov.cdc.prime.router.fhirengine.translation.hl7
+
+import ca.uhn.hl7v2.model.Type
+import ca.uhn.hl7v2.util.Terser
+import gov.cdc.prime.router.common.StringUtilities.trimAndTruncate
+import gov.cdc.prime.router.fhirengine.translation.hl7.utils.HL7Constants
+import gov.cdc.prime.router.fhirengine.translation.hl7.utils.HL7Constants.HD_FIELDS_LOCAL
+import gov.cdc.prime.router.fhirengine.translation.hl7.utils.HL7Constants.HD_TRUNCATION_LIMIT
+import gov.cdc.prime.router.fhirengine.translation.hl7.utils.HL7Utils
+import kotlin.math.min
+
+class HL7Truncator {
+
+    /**
+     * Trim and truncate the [value] according to the rules in [hl7Config] for [hl7Field].
+     * [terser] provides hl7 standards
+     */
+    fun trimAndTruncateValue(
+        value: String,
+        hl7Field: String,
+        terser: Terser,
+        truncateHDNamespaceIds: Boolean,
+        truncateHl7Fields: List<String>,
+        customLengthHl7Fields: Map<String, Int> = emptyMap(),
+    ): String {
+        val maxLength = getMaxLength(
+            hl7Field,
+            value,
+            terser,
+            truncateHDNamespaceIds,
+            truncateHl7Fields,
+            customLengthHl7Fields
+        )
+        return value.trimAndTruncate(maxLength)
+    }
+
+    /**
+     * Calculate for [hl7Field] and [value] the length to truncate the value according to the
+     * truncation rules in [hl7Config]. The [terser] is used to determine the HL7 specification length.
+     */
+    fun getMaxLength(
+        hl7Field: String,
+        value: String,
+        terser: Terser,
+        truncateHDNamespaceIds: Boolean,
+        truncateHl7Fields: List<String>,
+        customLengthHl7Fields: Map<String, Int> = emptyMap(),
+    ): Int? {
+        // The & character in HL7 is a sub sub field separator. A validly
+        // produced HL7 message should escape & characters as \T\ so that
+        // the HL7 parser doesn't interpret these as sub sub field separators.
+        // Because of this reason, all string values should go through the getTruncationLimitWithEncoding
+        // so that string values that contain sub sub field separators (^&~) will be properly truncated.
+        return when {
+            // This special case takes into account special rules needed by jurisdiction
+            truncateHDNamespaceIds && hl7Field in HD_FIELDS_LOCAL -> {
+                getTruncationLimitWithEncoding(value, HD_TRUNCATION_LIMIT)
+            }
+            hl7Field in customLengthHl7Fields -> {
```

Comment:
Added this block for the possibility of custom lengths in the future. It was a simple add and for the covid pipeline will be skipped over entirely.